src/NServiceBus.Core/Unicast/MessageHandlerRegistry.cs

```diff
@@ -61,152 +62,104 @@ public void RegisterHandler(Type handlerType)
+                MessageType = messageType.TypeHandle,
                 MethodDelegate = handleMethod
             };
-            List<DelegateHolder> methodList;
-            if (cache.TryGetValue(handler.TypeHandle, out methodList))
-            {
-                if (methodList.Any(x => x.MessageType == messageType))
-                {
-                    return;
-                }
-                methodList.Add(delegateHolder);
-            }
-            else
-            {
-                cache[handler.TypeHandle] = new List<DelegateHolder>
-                {
-                    delegateHolder
-                };
-            }
+            methodList.Add(delegateHolder);
         }
 
         static Action<object, object> GetMethod(Type targetType, Type messageType, Type interfaceGenericType)
         {
             var interfaceType = interfaceGenericType.MakeGenericType(messageType);
 
-            if (interfaceType.IsAssignableFrom(targetType))
+            if (!interfaceType.IsAssignableFrom(targetType))
             {
-                var methodInfo = targetType.GetInterfaceMap(interfaceType).TargetMethods.FirstOrDefault();
-                if (methodInfo != null)
-                {
-                    var target = Expression.Parameter(typeof(object));
-                    var param = Expression.Parameter(typeof(object));
+                return null;
+            }
 
-                    var castTarget = Expression.Convert(target, targetType);
-                    var castParam = Expression.Convert(param, methodInfo.GetParameters().First().ParameterType);
-                    var execute = Expression.Call(castTarget, methodInfo, castParam);
-                    return Expression.Lambda<Action<object, object>>(execute, target, param).Compile();
-                }
+            var methodInfo = targetType.GetInterfaceMap(interfaceType).TargetMethods.FirstOrDefault();
+            if (methodInfo == null)
+            {
+                return null;
             }
 
-            return null;
+            var target = Expression.Parameter(typeof(object));
+            var messageParam = Expression.Parameter(typeof(object));
+
+            var castTarget = Expression.Convert(target, targetType);
+
+            var methodParameters = methodInfo.GetParameters();
+            var messageCastParam = Expression.Convert(messageParam, methodParameters.ElementAt(0).ParameterType);
+
+            //var contextParameter = methodParameters.ElementAtOrDefault(1);
+            //if (IsNewContextApi(contextParameter))
+            //{
+            //    var contextType = contextParameter.ParameterType;
+            //    var contextCastParam = Expression.Convert(contextParam, contextType);
+            //    var execute = Expression.Call(castTarget, methodInfo, messageCastParam, contextCastParam);
+            //    return Expression.Lambda<Action<object, object, object>>(execute, target, messageParam, contextParam).Compile();
+            //}
```

Comment:
it kept it as reminder, since we will need similar wrapping when switching the timeout interfaces while handler interfaces stay sync. Will be removed soon.